cv32e40s/tests/programs/custom/debug_priv_test/debug_priv_test.c

```diff
@@ -0,0 +1,260 @@
+    __asm__ volatile("csrrw x0, mstatus, %0" :: "r"(MSTATUS_STD_VAL)); // set machine mode 
+    increment_mepc();
+}
+
+// Declaration of the debug struct 
+typedef union {
+  struct {
+    unsigned int start_delay      : 15; // 14: 0
+    unsigned int rand_start_delay : 1;  //    15
+    unsigned int pulse_width      : 13; // 28:16
+    unsigned int rand_pulse_width : 1;  //    29
+    unsigned int pulse_mode       : 1;  //    30    0 = level, 1 = pulse
+    unsigned int value            : 1;  //    31
+  } fields;
+  unsigned int bits;
+}  debug_req_control_t;
+
+/*
+When called it starts debug_mode
+*/
+void run_debug_mode(void) {
+  // debug control struct 
+  debug_req_control_t debug_req_control;
+  debug_req_control = (debug_req_control_t) {
+    .fields.value            = 1,
+    .fields.pulse_mode       = 1, //PULSE Mode
+    .fields.rand_pulse_width = 0,
+    .fields.pulse_width      = 5,// FIXME: BUG: one clock pulse cause core to lock up
+    .fields.rand_start_delay = 0,
+    .fields.start_delay      = 200
+  };
+
+  // this will initiate debug mode 
+  DEBUG_REQ_CONTROL_REG = debug_req_control.bits; 
+  // waits for debug to finish
+  while(glb_debug_status != 1){
+      ;
+  }
+  // resets the debug status bit
+  glb_debug_status = 0;
+}
+
+int main(void){// TODO: test will failed until issue #278 in core-v-verif/cv32e40s is fixed
+  setup_pmp();
+
+
+
+
+  // Test start:
+  /*
+  Have dcsr.ebreaku=0, be in U-mode, execute ebreak, ensure "normal" ebreak behavior and no debug entry.
+  */
+  set_u_mode();
+  asm volatile("ebreak");
+  assert_or_die(wmcause, 0x3, "Error: Illegal 'ebreak' did not trigger breakpoint exception!");
+
+
+
+  // TEST WILL FAIL UNTIL ISSUE #277 (openHW/cv32e40s) IS RESOLVED 
+  // TODO: comment the test back in when #277 has been resolved
+  // Test start:
+  /*
+  Transition out of D-mode (dret) into U-mode, while mstatus.mprv=1, ensure that when execution continues outside D-mode that mstatus.mprv=0.
+  */
+
+  /* 
+  glb_setmprv_test = 1; // flag the MRPRV-test for the debug module.
+  run_debug_mode();
+  asm volatile("ecall");
+  int mprvfield = get_field(wmstatus, MPRV_BIT, MPRV_BIT);
+  assert_or_die(mprvfield, 0x0, "Error: MPRV did not change to 0 after Debug --> User mode change! "); // check that MPRV = 0 after debug exit.
+  glb_setmprv_test = 0;
+ */
+
+
+
+
+  // Test start:
+  /* 
+  Transition out of D-mode, ensure that executions starts in the same privilege mode as was indicated in dcsr.prv (dcsr.prv=M and dcsr.prv=U).
+  */
+  // First check machine mode.
+  glb_check_prv_test = 1;
+  run_debug_mode();
+  asm volatile("ecall");
+  uint32_t check_MPP_bit = get_field(wmstatus, MPP_FIELD[0], MPP_FIELD[1]);
+  uint32_t check_prv_bit = get_field(glb_check_prv_test_val, PRV_FIELD[0], PRV_FIELD[1]);
+  assert_or_die(check_MPP_bit, check_prv_bit, "Error: previous privilege mode did not match the one set in dcsr.prv before exiting debug mode!\n");
```

Comment:
Could we add a check that the mode is indeed M/U as expected? Just to be safe.

Reply:
Does this trust that "dcsr.prv" was set to M/U? And does the first test trust that preceding tests left the mode in M?
It is probably correct, but it seems like a place where a small slip could cause the test to do something else than what we think it does.

Reply:
Both of the preceeding tests trap and the trap handler always puts the privilege mode back to machine mode. Therefore, I believe it is safe to assume that the privilege mode is machine mode at the start of the test in question. The test does not trust dcsr.prv functionality, it first sets dcsr.prv to the intended mode (M or U) and then once debug mode is finished it immediately does an ecall where the trap handler records mstatus. I believe that mstatus would then have to hold the immediate privilege mode that runs when coming out of debug mode and right before the ecall. If the dcsr.prv and mstatus privilege modes do not match then I believe that means the dcsr.prv is not functioning correctly. However, the test passes so then the assertions must also have passed for lines 198 and 207.